### PR TITLE
Add AppFramework compression middleware to gzip responses

### PR DESCRIPTION
**Pull request by rullzer**

### CODE REVIEW

File: lib/private/AppFramework/Middleware/GZipMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+			$this->useGZip = true;
+		}
+	}
+
+	public function afterException($controller, $methodName, \Exception $exception) {
+		$this->useGZip = false;
+	}
+
+
+	public function afterController($controller, $methodName, Response $response) {
```
Comment (rullzer) [inline lib/private/AppFramework/Middleware/GZipMiddleware.php:57]: Note to self. We should narrow this further I think.

basically OCS responses are good
JSON responses
or plain text responses

**Review by nickvergessen — APPROVED** [review]

File: lib/private/AppFramework/Middleware/CompressionMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		}
+
+		// Reset the state so we can check for the response types
+		$this->useGZip = false;
+		if ($response instanceof BaseResponse) {
+			$this->useGZip = true;
+		}
+		if ($response instanceof JSONResponse) {
+			$this->useGZip = true;
+		}
```
Comment (ChristophWurst) [inline lib/private/AppFramework/Middleware/CompressionMiddleware.php:76]: nit to remove the number of assignments

```php
$this->useGzip = ($response instanceof BaseResponse) || ($response instanceof JSONResponse);
```

**Review by ChristophWurst — APPROVED** [review]

File: lib/private/AppFramework/Middleware/CompressionMiddleware.php (hunk 2 above)
Comment (rullzer) [reply lib/private/AppFramework/Middleware/CompressionMiddleware.php:76]: I thought about that. But. I guess this list will grow... Like we could also chose to allow this for dataresponses that are plain text etc.

Comment (stefan-niedermann) [reply lib/private/AppFramework/Middleware/CompressionMiddleware.php:76]: Wouldn't it be possible to go for the mimetype? Like `if mimetype starts with "text/" use Gzip?`?

Comment (rullzer) [reply lib/private/AppFramework/Middleware/CompressionMiddleware.php:76]: sure that is what I mean with extentions. But first lets do this for OCS (as OCS is text by default). Then we can move on to others later :D

**Review by kesselb — APPROVED** [review]

**Comment by rullzer** [comment]
Added some tests. Just to be sure :)

**Comment by rullzer** [comment]
Note to self. This should not gzip when the status is 304 or 207... Actually maybe it should only try to gzip on a 200 response...

**Comment by rullzer** [comment]
> Note to self. This should not gzip when the status is 304 or 207... Actually maybe it should only try to gzip on a 200 response...

Ok made sure it will only do it for 200 responses. (Also for the followup 304 middleware).

File: lib/private/AppFramework/Middleware/CompressionMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			return $response;
+		}
+
+		// If we don't want to use gzip then we should not do that
+		if (!$this->useGZip) {
+			return $response;
+		}
+
+		// Reset the state so we can check for the response types
+		$this->useGZip = false;
```
Comment (MorrisJobke) [inline lib/private/AppFramework/Middleware/CompressionMiddleware.php:69]: Shouldn't we then just name the parameter different. There is no real world performance improvement to reuse the variable here. Just name it "allowGZip" or something like that for the check for the gzip header. Then the code is cleaner.

Comment (rullzer) [reply lib/private/AppFramework/Middleware/CompressionMiddleware.php:69]: sure. I guess this came from me rewriting the code a bunch of times :)

Comment (rullzer) [reply lib/private/AppFramework/Middleware/CompressionMiddleware.php:69]: actually thinking more. We don't even have to do the check in the before method...

File: .htaccess [hunk 4]
```diff
@@ -84,3 +84,7 @@ Options -Indexes
 <IfModule pagespeed_module>
   ModPagespeed Off
 </IfModule>
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
+
+ErrorDocument 403 /
```
Comment (kesselb) [inline .htaccess:89]: :thinking:

Comment (rullzer) [reply .htaccess]: My bad
We really need a precommit hook to see if the .htaccess changed

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Pull request merged**